Lib/inspect.py

```diff
@@ -818,11 +818,12 @@ def getfile(object):
         raise TypeError('{!r} is a built-in module'.format(object))
     if isclass(object):
         if hasattr(object, '__module__'):
+            # Protect against fetching the wrong source in PyREPL
+            if object.__module__ == '__main__':
+                raise OSError('source code not available')
```

Comment:
Hmm, I'm worried about this being a breaking change. `__module__` now has more precedent over `__file__`--maybe people could have been relying on that?

Reply:
Yeah I share similar concerns. Also why this only affects pyREPL and not any other __main__ module ?

Reply:
> Also why this only affects pyREPL and not any other **main** module ?

I think because for other `__main__` modules, `__main__.__file__` points to the correct source code.

In basic REPL, `sys.modules["__main__"]` doesn't have a `__file__` attribute, so it raises `OSError('source code not available')` when inspected. 

But with PyREPL, `sys.modules["__main__"].__file__` is `_pyrepl.__main__`, which is where the source code is being pulled from. So we could instead skip loading from `__file__` if we detect PyREPL is in use, what do you think?

Reply:
Wait, we could just delete `__file__` from `__main__`, right?
```diff
diff --git a/Lib/_pyrepl/main.py b/Lib/_pyrepl/main.py
index a6f824dcc4a..17c55eb9c9f 100644
--- a/Lib/_pyrepl/main.py
+++ b/Lib/_pyrepl/main.py
@@ -35,6 +35,7 @@ def interactive_console(mainmodule=None, quiet=False, pythonstartup=False):
         import __main__
         namespace = __main__.__dict__
         namespace.pop("__pyrepl_interactive_console", None)
+        namespace.pop("__file__", None)

     # sys._baserepl() above does this internally, we do it here
     startup_path = os.getenv("PYTHONSTARTUP")
```

Reply:
That fixes the problem no?

Reply:
Yeah, I think that would be much less invasive.